core/dbt/parser/schemas.py

```diff
@@ -114,29 +116,28 @@ class ParserRef:
     def __init__(self):
         self.column_info: Dict[str, ColumnInfo] = {}
 
-    def add(
-        self,
-        column: Union[HasDocs, UnparsedColumn],
-        description: str,
-        data_type: Optional[str],
-        constraints: Optional[List[str]],
-        constraints_check: Optional[str],
-        meta: Dict[str, Any],
-    ):
+    def _add(self, column: HasColumnProps):
         tags: List[str] = []
         tags.extend(getattr(column, "tags", ()))
         quote: Optional[bool]
         if isinstance(column, UnparsedColumn):
             quote = column.quote
         else:
             quote = None
+
+        if any(
+            c
+            for c in column.constraints
+            if not c["type"] or not ConstraintType.is_valid(c["type"])
```

Comment:
Along the same lines as @gshank's argument below, we should be validating the `constraint.type` when we set it, not when we use it. Once we get to this point, we should be able to assume that `type` exists for each `constraint`, and is always a valid type.

Reply:
This is the spot @gshank suggested validating, if I understood her correctly. I think validating earlier would require doing what I attempted in an earlier revision: Adding more structure to the `constraints` column on HasColumnProps. I encountered an issue with that approach, which is that it is serialized by hologram for partial parsing, which can't deal with enumerations.

Reply:
In general it's not very clean to handle enums in classes that represent yaml dictionaries (which is what most of the classes in unparsed.py are) since we need classes that will take the dictionary of strings that we get from yaml. So the validation has to be done in between the yaml-representing classes and instantiating nodes/configs. Somewhere.

In theory we could validate the dictionaries, and we do that in some places, but that's really early in parsing and unfortunately we don't distinguish between a bunch of overlapping nodes there, so in order to do that we'd have to separate out some parsing methods which are now common to a bunch of different nodes.

Reply:
At some point we should look at all of the places where validation is done and regularize it all and do it in a methodical way.

Reply:
Thanks @gshank for all of that context. That makes sense now.